### PR TITLE
Bring back the duplicates filter due to shifting completion item indexes after deduplication (continuation of fixes for #111)

### PR DESCRIPTION
**Pull request by dmitmel**
Alright alright, I realize that you are probably tired of #111 already, but I have found a pretty serious bug related to deduplication, which you removed in commit e599f0825cc6de637a70aee9b941698417d27768. As an example, let's say that the following completion list has to be displayed:

```lua
{
  {
    abbr = "print(...)~",
    dup = 1,
    empty = 1,
    equal = 1,
    kind = "Snippet",
    user_data = { cmp = 1911 },
    word = "print"
  }, {
    abbr = "print",
    dup = 0,
    empty = 1,
    equal = 1,
    kind = "Text",
    user_data = { cmp = 1709 },
    word = "print"
  }, {
    abbr = "printf(...)~",
    dup = 1,
    empty = 1,
    equal = 1,
    kind = "Snippet",
    user_data = { cmp = 1912 },
    word = "printf"
  }, {
    abbr = "vim_print",
    dup = 0,
    empty = 1,
    equal = 1,
    kind = "Text",
    user_data = { cmp = 1126 },
    word = "vim_print"
  }
}
```

The first three entries come from the Lua language server (`sumneko_lua`), the last one is from the cmp-buffer source. Now, after nvim-cmp calls `vim.fn.complete()` which performs deduplication internally, only `print(...)~`, `printf(...)~` and `vim_print` will be displayed to the user. However, when I select `vim_print`, its index (returned by `vim.fn.complete_info().selected`) will be 2 and not 3, but nvim-cmp will access the entry with index 2, will get the `printf` one and try to expand it instead of the `vim_print` which I have requested.

So basically, as a solution we must deduplicate on the Lua side, store the results in a table and then we will be safe to access that table with indices returned by `vim.fn.complete_info().selected`. I re-used most of the code from the commit e599f0825cc6de637a70aee9b941698417d27768, but changed the logic to use `item.word` instead of `item.abbr` because that's what Vim does. Also, as a sidenote, our implementation might very well be faster on long lists than Neovim's ([code](https://github.com/neovim/neovim/blob/5f8518b3f04c2295be24912dba78d3e34ca74b91/src/nvim/edit.c#L2352-L2366)) because they use linear lookup and we have a hashtable!

### CODE REVIEW

**Comment by hrsh7th** [comment]
LGTM 👍🏼
